### PR TITLE
[TORCH] Add support for aten.index_add Op via torch decomposition

### PR DESCRIPTION
**Pull request by AiyyappanMR**
This PR adds decomposition support for `aten.index_add`, which adds slices of a source tensor into a self tensor at positions given by a 1D index tensor, along a specified dimension, scaled by an optional alpha:

* `aten.scatter_add` requires an index tensor with the same rank and shape as `source`, while `aten.index_add`'s `index` is 1D (length equal to `source.size(dim)`). To bridge this, `index` is reshaped (`aten.view`) to size 1 in every dimension except `dim`, where it keeps its original length, then broadcast (`aten.expand`) out to `source`'s full shape.
* `source` is pre-scaled by `alpha` via `aten.mul.Scalar` before scattering, so the decomposition always routes through `aten.scatter_add(self, dim, expanded_index, scaled_source)`.
* This decomposition only fires when `dim` is a compile-time constant, since the reshape target shape needs `dim`'s position known statically to build the size list.

Added e2e tests: `IndexAdd1DFloatBasicModule_basic`, `IndexAdd2DFloatBasicModule_basic`, `IndexAdd2DFloatDim1Module_basic`, `IndexAdd1DFloatWithAlphaModule_basic`, `IndexAdd1DIntBasicModule_basic`, `IndexAdd3DIntBasicModule_basic`, `IndexAdd4DIntBasicModule_basic`.

### CODE REVIEW

**Comment by AiyyappanMR** [comment]
Would be great if someone could review this — @penguin-wwy / @sahas3 / @zjgarvey

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 1]
```diff
@@ -12501,6 +12501,94 @@ class DecomposeAtenIndexPutLikeOp
...
+    Type expandedIndexType = indexType.getWithSizesAndDtype(
+        sourceType.getSizes(), indexType.getOptionalDtype());
+    Value falseVal = Torch::ConstantBoolOp::create(rewriter, loc, false);
+    Value indexExpanded =
+        AtenExpandOp::create(rewriter, loc, expandedIndexType, indexReshaped,
+                             expandSizesList, falseVal);
+
+    // Perform the scatter_add operation.
+    Value result = AtenScatterAddOp::create(rewriter, loc, selfType, self,
+                                            op.getDim(), indexExpanded, source);
```
Comment (penguin-wwy) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Should `dim` be used instead of `op.getDim()`, since the former uses `toPositiveDim` to ensure non-negativity?

Comment (AiyyappanMR) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Good catch, thanks. Fixed by using the normalized dim value instead of op.getDim().

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 2]
```diff
@@ -12478,6 +12478,97 @@ class DecomposeAtenIndexPutLikeOp
 };
 } // namespace
 
+namespace {
+// Decompose aten.index_add into aten.scatter_add.
```
Comment (sahas3) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:12482]: Thanks for the enhancement, but I am curious if you considered leveraging PyTorch's decomposition for this op? 

The C++ decomp is beneficial if the same op appears from the ONNX frontend, since PyTorch decomp won't work there. Otherwise, we can decompose the op in PyTorch to avoid this additional code maintenance.

Comment (AiyyappanMR) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:12482]: Thanks, I didn't know that the importance of decomposing via MLIR (C++ decomp) is specifically for ops that are common to both the Torch and ONNX frontends, so it's shared across both paths. 

I've verified now that this op is only available on the Torch side, so I'll go ahead and proceed with decomposition via PyTorch instead. Will update the PR shortly.

Comment (AiyyappanMR) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:12482]: Sorry for the late reply, I was out of station.

Turns out `aten.index_add` is already handled by PyTorch’s `core_aten_decompositions()`, so it gets decomposed automatically in the `torch.export`/FX path. No C++ decomp is needed.

Verified this by tracing a model through export and confirming that `aten.index_add` doesn’t remain in the lowered IR.

Since this is Torch-only and not related to ONNX, I’ll close this out. Thanks for the pointer!
